klippy/extras/sgp40.py

```diff
@@ -0,0 +1,872 @@
+# SGP40 I2C-based humiditure sensor support
+#
+# Copyright (C) 2024 Stefan Dej <meteyou@gmail.com>
+# Based on the Nevermore sgp40.py code:
+# https://github.com/nevermore3d/Nevermore_Max/blob/master/Software/Klipper/
+#
+# This file may be distributed under the terms of the GNU GPLv3 license.
+#
+# The `VOCAlgorithm` class is based on the following work:
+# SPDX-FileCopyrightText:
+# Copyright (c) 2010 DFRobot Co.Ltd (http://www.dfrobot.com)
+# SPDX-License-Identifier: MIT
+# Class and algorithm to convert Sensirion sgp40 raw reading to indexed voc
+# readings.
+# Author(s): yangfeng
+
+import logging
+import math
+from . import bus
+from struct import unpack_from
+
+SGP40_I2C_ADDR = 0x59
+SGP40_WORD_LEN = 2
+
+SGP40_CMD = {
+    "GET_SERIAL": [0x36, 0x82],
+    "SOFT_RESET": [0x00, 0x06],
+    "SELF_TEST": [0x28, 0x0E],
+    "MEASURE_RAW_NO_COMP": [0x26, 0x0F, 0x80, 0x00, 0xA2, 0x66, 0x66,
+                             0x93],
+    "MEASURE_RAW": [0x26, 0x0F]
+}
+
+# Constants for the VOC algorithm
+_VOC_SAMPLING_INTERVAL = 1
+_VOC_INITIAL_BLACKOUT = 45
+_VOC_INDEX_GAIN = 230
+_VOC_SRAW_STD_INITIAL = 50
+_VOC_SRAW_STD_BONUS = 220
+_VOC_TAU_MEAN_VARIANCE_HOURS = 12
+_VOC_TAU_INITIAL_MEAN = 20
+_VOC_INITI_DURATION_MEAN = 2700
+_VOC_INITI_TRANSITION_MEAN = 0.01
+_VOC_TAU_INITIAL_VARIANCE = 2500
+_VOC_INITI_DURATION_VARIANCE = 5220
+_VOC_INITI_TRANSITION_VARIANCE = 0.01
+_VOC_GATING_THRESHOLD = 340
+_VOC_GATING_THRESHOLD_INITIAL = 510
+_VOC_GATING_THRESHOLD_TRANSITION = 0.09
+_VOC_GATING_MAX_DURATION_MINUTES = 180
+_VOC_GATING_MAX_RATIO = 0.3
+_VOC_SIGMOID_L = 500
+_VOC_SIGMOID_K = -0.0065
+_VOC_SIGMOID_X0 = 213
+_VOC_INDEX_OFFSET_DEFAULT = 100
+_VOC_LP_TAU_FAST = 20
+_VOC_LP_TAU_SLOW = 500
+_VOC_LP_ALPHA = -0.2
+_VOC_PERSISTENCE_UPTIME_GAMMA = 10800
+_VOC_MEAN_VARIANCE_ESTIMATOR__GAMMA_SCALING = 64
+_VOC_MEAN_VARIANCE_ESTIMATOR__FIX16_MAX = 32767
+_FIX16_MAXIMUM = 0x7FFFFFFF
+_FIX16_MINIMUM = 0x80000000
+_FIX16_OVERFLOW = 0x80000000
+_FIX16_ONE = 0x00010000
+
+class SGP40:
+    def __init__(self, config):
+        self.printer = config.get_printer()
+        self.name = config.get_name().split()[-1]
+        self.reactor = self.printer.get_reactor()
+        self.i2c = bus.MCU_I2C_from_config(
+            config, default_addr=SGP40_I2C_ADDR, default_speed=100000)
+        self.report_time = config.getint('sgp40_report_time', 1, minval=1)
```

Comment:
sampling must be at 1Hz for this algorithm according to the datasheet. The algorithm parameters all assume such. It should not be configurable.